src/main/java/org/broadinstitute/hellbender/tools/spark/utils/FlatMapGluer.java

```diff
@@ -0,0 +1,62 @@
+package org.broadinstitute.hellbender.tools.spark.utils;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+public class FlatMapGluer<I,O> implements Iterator<O> {
+    private final Function<I,Iterator<O>> flatMapFunc;
+    private final Iterator<? extends I> inputIterator;
+    private I sentinel;
```

Comment:
Actually, any subtype of I can be passed as an I (because it _IsA_ I).  So your example of a special subclass will work fine.
It's just the type of the collection that needs to be covariant (because an Iterator<J> is not in an _IsA_ relationship with Iterator<I>, even if J extends I).
Java generics are big, big fun!